File: src/core/fuzzy.ts
// fzf-v1-style fuzzy subsequence scorer. Pure core — reused by the New + Resume filters
// (and mirrored ~40 LOC in the Swift GUI). Returns matched positions so callers can highlight.
// No substring pre-filter anywhere: that would exclude exactly the gapped matches this exists to surface.

export interface FuzzyResult {
  score: number;
  positions: number[];
}

const BONUS_CONSECUTIVE = 8; // contiguous runs beat scattered boundary hits
const BONUS_BOUNDARY = 6; // match at a word boundary (start / after separator / camelCase)
const SEP = /[/\\_\-. ]/;

/**
 * Score `text` against `query` (case-insensitive subsequence). Higher = better.
 * Returns null if `query` is not a subsequence of `text`. Empty query → score 0, no positions.
 */
export function fuzzyMatch(query: string, text: string): FuzzyResult | null {
  if (!query) return { score: 0, positions: [] };
  const q = query.toLowerCase();
  const t = text.toLowerCase();
  const positions: number[] = [];
  let qi = 0;
  let score = 0;
  let prevMatch = -2;

  for (let ti = 0; ti < t.length && qi < q.length; ti++) {
    if (t[ti] !== q[qi]) continue;
    let bonus = 1;
    if (prevMatch === ti - 1) bonus += BONUS_CONSECUTIVE;
    const prev = ti > 0 ? text[ti - 1] : '';
    if (ti === 0 || SEP.test(prev)) {
      bonus += BONUS_BOUNDARY;
    } else if (prev && prev === prev.toLowerCase() && text[ti] !== text[ti].toLowerCase()) {
      bonus += BONUS_BOUNDARY - 2; // camelCase boundary (lower→Upper)
    }
    score += bonus;
    positions.push(ti);
    prevMatch = ti;
    qi++;
  }

  if (qi < q.length) return null; // not all query chars consumed
  // Prefer earlier first match and tighter spans.
  const span = positions[positions.length - 1] - positions[0];
  score -= positions[0] * 0.5 + span * 0.2;
  return { score, positions };
}

/** Rank items by fuzzy score against `getText(item)`, dropping non-matches. Stable for empty query. */
export function fuzzyRank<T>(query: string, items: T[], getText: (item: T) => string): Array<{ item: T; result: FuzzyResult }> {
  if (!query) return items.map((item) => ({ item, result: { score: 0, positions: [] } }));
  const scored: Array<{ item: T; result: FuzzyResult; i: number }> = [];
  items.forEach((item, i) => {
    const result = fuzzyMatch(query, getText(item));
    if (result) scored.push({ item, result, i });
  });
  scored.sort((a, b) => (b.result.score - a.result.score) || (a.i - b.i)); // stable on ties
  return scored.map(({ item, result }) => ({ item, result }));
}

File: gui/AgentCliMenuBar/Sources/AgentCliMenuBar/Fuzzy.swift
import Foundation

// fzf-v1-style fuzzy subsequence scorer — a Swift port of src/core/fuzzy.ts.
// Same weights, same behavior, so the GUI ranks New dirs + Resume sessions exactly
// like the terminal does. No substring pre-filter: gapped matches are the point.

struct FuzzyResult {
    let score: Double
    let positions: [Int]
}

enum Fuzzy {
    private static let bonusConsecutive = 8.0   // contiguous runs beat scattered boundary hits
    private static let bonusBoundary = 6.0       // match at a word boundary (start / after separator / camelCase)
    private static let separators: Set<Character> = ["/", "\\", "_", "-", ".", " "]

    /// Score `text` against `query` (case-insensitive subsequence). Higher = better.
    /// Returns nil if `query` is not a subsequence. Empty query → score 0.
    static func match(_ query: String, _ text: String) -> FuzzyResult? {
        if query.isEmpty { return FuzzyResult(score: 0, positions: []) }
        let q = Array(query.lowercased())
        let original = Array(text)
        let t = Array(text.lowercased())
        var positions: [Int] = []
        var qi = 0
        var score = 0.0
        var prevMatch = -2
        var ti = 0
        while ti < t.count && qi < q.count {
            if t[ti] != q[qi] { ti += 1; continue }
            var bonus = 1.0
            if prevMatch == ti - 1 { bonus += bonusConsecutive }
            let prev: Character? = ti > 0 ? original[ti - 1] : nil
            if ti == 0 || (prev != nil && separators.contains(prev!)) {
                bonus += bonusBoundary
            } else if let p = prev,
                      String(p) == String(p).lowercased(),       // prev unchanged by lowercasing (matches TS)
                      String(original[ti]) != String(original[ti]).lowercased() { // cur is uppercase
                bonus += bonusBoundary - 2 // camelCase boundary (lower/digit/sym → Upper)
            }
            score += bonus
            positions.append(ti)
            prevMatch = ti
            qi += 1
            ti += 1
        }
        if qi < q.count { return nil } // not all query chars consumed
        let span = positions[positions.count - 1] - positions[0]
        score -= Double(positions[0]) * 0.5 + Double(span) * 0.2
        return FuzzyResult(score: score, positions: positions)
    }

    /// Rank items by score against `getText`, dropping non-matches. Empty query keeps original order (stable).
    static func rank<T>(_ query: String, _ items: [T], _ getText: (T) -> String) -> [T] {
        if query.isEmpty { return items }
        let scored: [(item: T, score: Double, i: Int)] = items.enumerated().compactMap { i, item in
            guard let r = match(query, getText(item)) else { return nil }
            return (item, r.score, i)
        }
        return scored
            .sorted { $0.score != $1.score ? $0.score > $1.score : $0.i < $1.i }
            .map { $0.item }
    }
}
